File: src/auth/controllers/auth.controller.js
// src/auth/controllers/auth.controller.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const { getUserByEmail, createUser, getUserRole, updateUser, deleteUser, updateUserTipoUsuario, getAllUsers } = require('../models/users.model');
const { getRoleByName, createRole, getAllRoles } = require('../models/roles.model');
const { getPermisosByRol } = require('../models/roles_permisos.model');

const JWT_SECRET = process.env.JWT_SECRET;
const JWT_EXPIRES_IN = process.env.JWT_EXPIRES_IN || '2h';

async function register(req, res) {
  try {
    const { nombre, email, password, tipo_usuario } = req.body;
    if (!nombre || !email || !password) return res.status(400).json({ message: 'Faltan datos' });

    const salt = await bcrypt.genSalt(10);
    const hash = await bcrypt.hash(password, salt);

    const user = await createUser({ nombre, email, tipo_usuario, passwordHash: hash });
    return res.status(201).json({ message: 'Usuario creado', user });
  } catch (err) {
    console.error('Error en register:', err);
    return res.status(400).json({ error: err.message || 'Error creando usuario' });
  }
}

async function login(req, res) {
  try {
    // Ahora solo email + password en el body
    const { email, password } = req.body;
    if (!email || !password) {
      return res.status(400).json({ message: 'Faltan datos: email y password son requeridos' });
    }

    const user = await getUserByEmail(email);
    if (!user) return res.status(401).json({ message: 'Credenciales inválidas' });

    const match = await bcrypt.compare(password, user.password_hash);
    if (!match) return res.status(401).json({ message: 'Credenciales inválidas' });

    // Verificar que el usuario esté activo
    if (typeof user.activo !== 'undefined' && !user.activo) {
      return res.status(401).json({ message: 'Su cuenta está desactivada. Contacte al administrador' });
    }

    // obtener único rol del usuario usando la FK tipo_usuario
    const role = await getUserRole(user.id_usuario);
    const roleNames = role ? [role.nombre_rol] : [];
    const roleIds = role ? [role.id_rol] : [];

    // obtener permisos a partir de roleIds (si existen)
    let permisos = [];
    for (const id_rol of roleIds) {
      const permisosData = await getPermisosByRol(id_rol);
      permisos = permisos.concat(permisosData.map(p => p.nombre_permiso));
    }
    // Normalizar permisos (lowercase) y eliminar duplicados
    permisos = [...new Set((permisos || []).map(p => String(p).toLowerCase()))];

    const roleNamesNormalized = (roleNames || []).map(r => String(r).toLowerCase());

    if (!JWT_SECRET) {
      console.error('FATAL: JWT_SECRET no configurado en .env');
      return res.status(500).json({ error: 'Configuración del servidor incompleta (JWT_SECRET).' });
    }

    const tokenPayload = { 
      id: user.id_usuario, 
      email: user.email, 
      roles: roleNamesNormalized, 
      roleIds,
      permisos
    };
    const token = jwt.sign(tokenPayload, JWT_SECRET, { expiresIn: JWT_EXPIRES_IN });

    return res.json({
      token,
      user: { 
        id: user.id_usuario, 
        nombre: user.nombre, 
        email: user.email, 
        roles: roleNamesNormalized,
        permisos
      }
    });
  } catch (err) {
    console.error('Error en login:', err);
    return res.status(500).json({ error: err.message || 'Error interno del servidor' });
  }
}

async function assignRole(req, res) {
  try {
    const { id_usuario } = req.params;
    const { nombre_rol, descripcion } = req.body;
    if (!nombre_rol) return res.status(400).json({ message: 'nombre_rol requerido' });

    let role = await getRoleByName(nombre_rol);
    if (!role) role = await createRole(nombre_rol, descripcion);

    const result = await updateUserTipoUsuario(parseInt(id_usuario, 10), role.id);
    if (!result || result.rowsAffected === 0) {
      return res.status(404).json({ message: 'Usuario no encontrado' });
    }
    return res.json({ message: 'Rol asignado' });
  } catch (err) {
    console.error('Error en assignRole:', err);
    return res.status(500).json({ error: err.message || 'Error asignando rol' });
  }
}

async function modifyUser(req, res) {
  try {
    const { id_usuario } = req.params;
    const id = Number(id_usuario);
    if (!Number.isInteger(id) || id <= 0) return res.status(400).json({ message: 'id_usuario inválido' });

    const { nombre, email, tipo_usuario, activo, password } = req.body;
    console.log('DEBUG modifyUser - id:', id, 'body:', req.body);

    const roles = req.user?.roles || [];
    const isAdmin = roles.map(r => String(r).toLowerCase()).includes('admin');
    if (!isAdmin && typeof tipo_usuario !== 'undefined') {
      return res.status(403).json({ message: 'No tienes permiso para cambiar el rol del usuario' });
    }

    let password_hash = null;
    if (password) {
      const salt = await bcrypt.genSalt(10);
      password_hash = await bcrypt.hash(password, salt);
    }

    const result = await updateUser(id, { nombre, email, tipo_usuario, activo, password_hash });

    if (!result || result.rowsAffected === 0) {
      return res.status(404).json({ message: 'Usuario no encontrado o sin cambios' });
    }
    return res.json({ message: 'Usuario actualizado' });
  } catch (err) {
    console.error('Error en modifyUser:', err);
    return res.status(500).json({ error: err.message || 'Error actualizando usuario' });
  }
}

async function removeUser(req, res) {
  try {
    const { id_usuario } = req.params;
    const id = Number(id_usuario);
    if (!Number.isInteger(id) || id <= 0) return res.status(400).json({ message: 'id_usuario inválido' });

    const roles = req.user?.roles || [];
    const isAdmin = roles.map(r => String(r).toLowerCase()).includes('admin');
    const requesterId = req.user?.id;
    if (!isAdmin && requesterId !== id) {
      return res.status(403).json({ message: 'No tienes permiso para eliminar este usuario' });
    }

    console.log('DEBUG removeUser - id:', id);
    const result = await deleteUser(id);

    if (!result || result.rowsAffected === 0) {
      return res.status(404).json({ message: 'Usuario no encontrado' });
    }
    return res.json({ message: 'Usuario eliminado' });
  } catch (err) {
    console.error('Error en removeUser:', err);
    if (err && (err.number === 547 || err.code === 'ERESTRICT' || err.code === '23503')) {
      return res.status(409).json({ error: 'No se puede eliminar usuario: existen registros relacionados.' });
    }
    return res.status(500).json({ error: err.message || 'Error eliminando usuario' });
  }
}

async function listUsers(req, res) {
  try {
    const users = await getAllUsers();
    return res.json({ users });
  } catch (err) {
    console.error('Error en listUsers:', err);
    return res.status(500).json({ error: err.message || 'Error listando usuarios' });
  }
}

async function listRoles(req, res) {
  try {
    const roles = await getAllRoles();
    return res.json({ roles });
  } catch (err) {
    console.error('Error en listRoles:', err);
    return res.status(500).json({ error: err.message || 'Error listando roles' });
  }
}

module.exports = { register, login, assignRole, modifyUser, removeUser, listRoles, listUsers };

File: src/auth/controllers/auth.controller.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.IdentityModel.Tokens;
using SmartERP.Models;
using SmartERP.Services;
using System;
using System.IdentityModel.Tokens.Jwt;
using System.Linq;
using System.Security.Claims;
using System.Text;
using System.Threading.Tasks;

namespace SmartERP.Controllers
{
[ApiController]
[Route("api/[controller]")]
public class AuthController : ControllerBase
{
private readonly IUserService _userService;
private readonly IRoleService _roleService;
private readonly IPermisoService _permisoService;
private readonly IConfiguration _config;

```
    public AuthController(
        IUserService userService,
        IRoleService roleService,
        IPermisoService permisoService,
        IConfiguration config)
    {
        _userService = userService;
        _roleService = roleService;
        _permisoService = permisoService;
        _config = config;
    }

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(request.Nombre) || string.IsNullOrWhiteSpace(request.Email) || string.IsNullOrWhiteSpace(request.Password))
                return BadRequest(new { message = "Faltan datos" });

            var existing = await _userService.GetByEmailAsync(request.Email);
            if (existing != null)
                return BadRequest(new { message = "El usuario ya existe" });

            string hash = BCrypt.Net.BCrypt.HashPassword(request.Password);

            var user = new Usuario
            {
                Nombre = request.Nombre,
                Email = request.Email,
                TipoUsuario = request.TipoUsuario,
                PasswordHash = hash
            };

            var created = await _userService.CreateAsync(user);
            return Created("", new { message = "Usuario creado", user = created });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en Register: {ex.Message}");
            return StatusCode(400, new { error = ex.Message });
        }
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(request.Email) || string.IsNullOrWhiteSpace(request.Password))
                return BadRequest(new { message = "Faltan datos: email y password son requeridos" });

            var user = await _userService.GetByEmailAsync(request.Email);
            if (user == null || !BCrypt.Net.BCrypt.Verify(request.Password, user.PasswordHash))
                return Unauthorized(new { message = "Credenciales inválidas" });

            if (user.Activo.HasValue && !user.Activo.Value)
                return Unauthorized(new { message = "Su cuenta está desactivada. Contacte al administrador" });

            var role = await _roleService.GetRoleByUserIdAsync(user.IdUsuario);
            var roleNames = role != null ? new[] { role.NombreRol.ToLower() } : Array.Empty<string>();
            var roleIds = role != null ? new[] { role.IdRol } : Array.Empty<int>();

            var permisos = await _permisoService.GetPermisosByRolesAsync(roleIds);
            var permisosList = permisos.Select(p => p.NombrePermiso.ToLower()).Distinct().ToArray();

            string secret = _config["Jwt:Secret"];
            string expiresIn = _config["Jwt:ExpiresIn"] ?? "2h";
            if (string.IsNullOrEmpty(secret))
                return StatusCode(500, new { error = "Configuración JWT_SECRET no encontrada." });

            var token = GenerateJwtToken(user, roleNames, roleIds, permisosList, secret, expiresIn);

            return Ok(new
            {
                token,
                user = new
                {
                    id = user.IdUsuario,
                    nombre = user.Nombre,
                    email = user.Email,
                    roles = roleNames,
                    permisos = permisosList
                }
            });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en Login: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    private string GenerateJwtToken(Usuario user, string[] roles, int[] roleIds, string[] permisos, string secret, string expiresIn)
    {
        var claims = new[]
        {
            new Claim(JwtRegisteredClaimNames.Sub, user.Email),
            new Claim("id", user.IdUsuario.ToString()),
            new Claim("roles", string.Join(",", roles)),
            new Claim("roleIds", string.Join(",", roleIds)),
            new Claim("permisos", string.Join(",", permisos))
        };

        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(secret));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var expiry = DateTime.Now.AddHours(2);

        var token = new JwtSecurityToken(
            issuer: "SmartERP",
            audience: "SmartERP",
            claims: claims,
            expires: expiry,
            signingCredentials: creds);

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    [HttpPost("assign-role/{id_usuario:int}")]
    public async Task<IActionResult> AssignRole(int id_usuario, [FromBody] AssignRoleRequest request)
    {
        try
        {
            if (string.IsNullOrWhiteSpace(request.NombreRol))
                return BadRequest(new { message = "nombre_rol requerido" });

            var role = await _roleService.GetRoleByNameAsync(request.NombreRol)
                ?? await _roleService.CreateRoleAsync(request.NombreRol, request.Descripcion);

            var updated = await _userService.UpdateTipoUsuarioAsync(id_usuario, role.IdRol);
            if (!updated)
                return NotFound(new { message = "Usuario no encontrado" });

            return Ok(new { message = "Rol asignado" });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en AssignRole: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpPut("{id_usuario:int}")]
    public async Task<IActionResult> ModifyUser(int id_usuario, [FromBody] UpdateUserRequest request)
    {
        try
        {
            if (id_usuario <= 0)
                return BadRequest(new { message = "id_usuario inválido" });

            var userRoles = HttpContext.User.FindFirst("roles")?.Value?.Split(',') ?? Array.Empty<string>();
            bool isAdmin = userRoles.Contains("admin");

            if (!isAdmin && request.TipoUsuario != null)
                return Forbid(new { message = "No tienes permiso para cambiar el rol del usuario" });

            string? passwordHash = null;
            if (!string.IsNullOrWhiteSpace(request.Password))
                passwordHash = BCrypt.Net.BCrypt.HashPassword(request.Password);

            bool updated = await _userService.UpdateUserAsync(id_usuario, request.Nombre, request.Email, request.TipoUsuario, request.Activo, passwordHash);

            if (!updated)
                return NotFound(new { message = "Usuario no encontrado o sin cambios" });

            return Ok(new { message = "Usuario actualizado" });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en ModifyUser: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpDelete("{id_usuario:int}")]
    public async Task<IActionResult> RemoveUser(int id_usuario)
    {
        try
        {
            var roles = HttpContext.User.FindFirst("roles")?.Value?.Split(',') ?? Array.Empty<string>();
            bool isAdmin = roles.Contains("admin");
            var requesterId = HttpContext.User.FindFirst("id")?.Value;

            if (!isAdmin && requesterId != id_usuario.ToString())
                return Forbid(new { message = "No tienes permiso para eliminar este usuario" });

            bool deleted = await _userService.DeleteUserAsync(id_usuario);
            if (!deleted)
                return NotFound(new { message = "Usuario no encontrado" });

            return Ok(new { message = "Usuario eliminado" });
        }
        catch (Exception ex)
        {
            if (ex.Message.Contains("FK") || ex.Message.Contains("23503"))
                return Conflict(new { error = "No se puede eliminar usuario: existen registros relacionados." });

            Console.WriteLine($"Error en RemoveUser: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpGet("users")]
    public async Task<IActionResult> ListUsers()
    {
        try
        {
            var users = await _userService.GetAllUsersAsync();
            return Ok(new { users });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en ListUsers: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }

    [HttpGet("roles")]
    public async Task<IActionResult> ListRoles()
    {
        try
        {
            var roles = await _roleService.GetAllRolesAsync();
            return Ok(new { roles });
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error en ListRoles: {ex.Message}");
            return StatusCode(500, new { error = ex.Message });
        }
    }
}

// DTOs
public record RegisterRequest(string Nombre, string Email, string Password, int? TipoUsuario);
public record LoginRequest(string Email, string Password);
public record AssignRoleRequest(string NombreRol, string? Descripcion);
public record UpdateUserRequest(string? Nombre, string? Email, int? TipoUsuario, bool? Activo, string? Password);
```

}
